File: holoeasy-core/src/main/java/org/holoeasy/hologram/HologramLoader.java
package org.holoeasy.hologram;

import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.holoeasy.line.Line;
import org.holoeasy.line.LineImpl;
import org.holoeasy.line.TextLine;

import java.util.Collections;
import java.util.List;
import java.util.function.Consumer;

public enum HologramLoader {

    SINGLETON(
        hologram -> {
            if (hologram.getLines().size() > 1) {
                throw new RuntimeException("Hologram '" + hologram.getId() + "' has more than 1 line.");
            }
            Location cloned = hologram.getLocation().clone();
            LineImpl<?> line = (LineImpl<?>) hologram.getLines().get(0);
            line.setCurrentLocation(cloned);
        },
        hologram -> {
            LineImpl<?> line = (LineImpl<?>) hologram.getLines().get(0);
            line.setCurrentLocation(hologram.getLocation().clone());
            for (Player player : hologram.getPvt().getSeeingPlayers()) {
                line.teleport(player);
            }
        }
    ),

    TEXT_SEQUENTIAL(
        hologram -> textSequential(hologram, true),
        hologram -> {
            textSequential(hologram, false);
            // TODO: Quando si teleporta, gli ologrammi diventano inaspettatamente distanti. Capire perché.
        }
    ),

    TEXT_BLOCK_STANDARD(
        hologram -> {
            final double LINE_HEIGHT = 0.28;

            Location hologramLocation = hologram.getLocation().clone();
            if (hologram.getLines().size() == 1) {
                LineImpl<?> line = (LineImpl<?>) hologram.getLines().get(0);
                line.setCurrentLocation(hologramLocation);
            } else {
                List<?> lines = new java.util.ArrayList<>(hologram.getLines());
                Collections.reverse(lines);
                hologram.getLines().clear();

                hologramLocation.subtract(0.0, LINE_HEIGHT, 0.0);

                for (int j = 0; j < lines.size(); j++) {
                    LineImpl<?> line = (LineImpl<?>) lines.get(j);
                    double up = LINE_HEIGHT;
                    if (j > 0) {
                        switch (((LineImpl<?>) lines.get(j - 1)).getType()) {
                            case ITEM_LINE: up = -1.5; break;
                            case BLOCK_LINE: up = -0.19; break;
                            case EXTERNAL: case TEXT_LINE: case CLICKABLE_TEXT_LINE: case DISPLAY_BLOCK_LINE: break;
                            case DISPLAY_TEXT_LINE: throw new UnsupportedOperationException();
                        }
                    }
                    switch (line.getType()) {
                        case TEXT_LINE:
                        case CLICKABLE_TEXT_LINE:
                        case BLOCK_LINE:
                            line.setCurrentLocation(hologramLocation.add(0.0, up, 0.0).clone());
                            hologram.getLines().add(0, line);
                            break;
                        case ITEM_LINE:
                            line.setCurrentLocation(hologramLocation.add(0.0, 0.6, 0.0).clone());
                            hologram.getLines().add(0, line);
                            break;
                        default:
                            throw new RuntimeException("This method load does not support line type " + line.getType().name());
                    }
                }
            }
        },
        hologram -> {
            final double LINE_HEIGHT = 0.28;

            List<Line<?>> lines = hologram.getLines();
            LineImpl<?> firstLine = (LineImpl<?>) lines.get(0);
            double baseY = firstLine.currentLocation().getY();
            double destY = hologram.getLocation().getY() - LINE_HEIGHT;
            switch (firstLine.getType()) {
                case TEXT_LINE:
                case CLICKABLE_TEXT_LINE:
                case BLOCK_LINE:
                    destY += LINE_HEIGHT;
                    break;
                default:
                    destY += 0.6;
            }
            teleportStandard(hologram, destY, firstLine);
            for (int j = 1; j < lines.size(); j++) {
                LineImpl<?> tempLine = (LineImpl<?>) lines.get(j);
                teleportStandard(
                    hologram,
                    destY + Math.abs(baseY - tempLine.currentLocation().getY()),
                    tempLine
                );
            }
        }
    );



    private final Consumer<Hologram> load;
    private final Consumer<Hologram> teleport;

    HologramLoader(Consumer<Hologram> load, Consumer<Hologram> teleport) {
        this.load = load;
        this.teleport = teleport;
    }

    public void load(Hologram hologram) {
        load.accept(hologram);
    }

    public void teleport(Hologram hologram) {
        teleport.accept(hologram);
    }

    private static void teleportStandard(Hologram hologram, double destY, LineImpl<?> tempLine) {
        Location dest = hologram.getLocation().clone();
        dest.setY(destY);
        tempLine.setCurrentLocation(dest);
        for (Player player : hologram.getPvt().getSeeingPlayers()) {
            tempLine.teleport(player);
        }
    }

    private static void textSequential(Hologram hologram, boolean add) {
        Location cloned = hologram.getLocation().clone();
        for (Line<?> line : hologram.getLines()) {
            switch (((LineImpl<?>) line).getType()) {
                case TEXT_LINE:
                case CLICKABLE_TEXT_LINE: {
                    TextLine tL = (TextLine) line;
                    tL.setCurrentLocation(cloned.clone());
                    if (add) {
                        // hologram.getLines().add(0, tL);
                    } else {
                        for (Player player : hologram.getPvt().getSeeingPlayers()) {
                            tL.teleport(player);
                        }
                    }
                    cloned.setZ(cloned.getZ() + 0.175 * tL.getValue().length());
                    break;
                }
                default:
                    throw new RuntimeException("This method load supports only TextLine & ClickableTextLine.");
            }
        }
    }
}

File: holoeasy-core/src/main/kotlin/org/holoeasy/hologram/HologramLoader.kt
package org.holoeasy.hologram


import org.holoeasy.line.LineImpl
import org.holoeasy.line.TextLine
import kotlin.math.abs

enum class HologramLoader(
    val load: (Hologram) -> Unit,
    val teleport: (Hologram) -> Unit
) {

    SINGLETON(load = { hologram ->
        if (hologram.lines.size > 1) {
            throw RuntimeException("Hologram '${hologram.id}' has more than 1 line.")
        }

        val cloned = hologram.location.clone()

        val line = hologram.lines[0] as LineImpl

        line.setCurrentLocation(cloned)
    }, teleport = { hologram ->
        val line = hologram.lines[0] as LineImpl

        line.setCurrentLocation(hologram.location.clone())
        hologram.pvt.seeingPlayers.forEach(line::teleport)
    }),

    TEXT_SEQUENTIAL(load = { hologram ->
        textSequential(hologram, true)

    }, teleport = { hologram ->
        textSequential(hologram, false)
        // TODO: When teleporting, the holograms unexpectedly become distant. Understand why.
    }),

    TEXT_BLOCK_STANDARD(load = { hologram ->
        val hologramLocation = hologram.location.clone()

        if (hologram.lines.size == 1) {
            val line = hologram.lines[0] as LineImpl

            line.setCurrentLocation(hologramLocation)
        } else {
            // reverse A - B - C to C - B - A
            val lines = hologram.lines.reversed()
            hologram.lines.clear()

            hologramLocation.subtract(0.0, LINE_HEIGHT, 0.0)

            for (j in lines.indices) {
                val line = lines[j] as LineImpl

                var up = LINE_HEIGHT

                if (j > 0) {
                    when (lines[j - 1].type) {
                        LineImpl.Type.ITEM_LINE -> up = -1.5
                        LineImpl.Type.BLOCK_LINE -> up = -0.19
                        LineImpl.Type.EXTERNAL -> {}
                        LineImpl.Type.TEXT_LINE -> {}
                        LineImpl.Type.CLICKABLE_TEXT_LINE -> {}
                        org.holoeasy.line.LineImpl.Type.DISPLAY_BLOCK_LINE -> {}
                    }
                }

                when (line.type) {
                    LineImpl.Type.TEXT_LINE, LineImpl.Type.CLICKABLE_TEXT_LINE, LineImpl.Type.BLOCK_LINE -> {
                        line.setCurrentLocation(hologramLocation.add(0.0, up, 0.0).clone())
                        hologram.lines.add(0, line)
                    }

                    LineImpl.Type.ITEM_LINE -> {
                        line.setCurrentLocation(hologramLocation.add(0.0, 0.6, 0.0).clone())
                        hologram.lines.add(0, line)
                    }


                    else -> throw RuntimeException("This method load does not support line type " + line.type.name)
                }
            }
        }


    }, teleport = { hologram ->
        val lines = hologram.lines
        val firstLine = lines[0] as LineImpl
        // Obtain the Y position of the first line and then calculate the distance to all lines to maintain this distance
        val baseY: Double =
            firstLine.currentLocation()?.y ?: throw RuntimeException("First line has not a location")
        // Get position Y where to teleport the first line
        var destY = (hologram.location.y - LINE_HEIGHT)

        destY += when (firstLine.type) {
            LineImpl.Type.TEXT_LINE, LineImpl.Type.CLICKABLE_TEXT_LINE, LineImpl.Type.BLOCK_LINE -> LINE_HEIGHT
            else -> 0.6
        }

        // Teleport the first line
        teleportStandard(hologram, destY, firstLine)
        var tempLine: LineImpl<*>
        for (j in 1 until lines.size) {
            tempLine = lines[j] as LineImpl
            /*
            Teleport from the second line onwards.
            The final height is found by adding to that of the first line the difference that was present when it was already spawned
            */
            teleportStandard(
                hologram, destY + abs(
                    baseY -
                            (tempLine.currentLocation()?.y
                                ?: throw RuntimeException("Missing location of line $tempLine"))
                ), tempLine
            )
        }
    }),


    ;
}


private const val LINE_HEIGHT = 0.28


private fun teleportStandard(hologram: Hologram, destY: Double, tempLine: LineImpl<*>) {
    val dest = hologram.location.clone()
    dest.y = destY
    tempLine.setCurrentLocation(dest)
    hologram.pvt.seeingPlayers.forEach(tempLine::teleport)
}


private fun textSequential(hologram: Hologram, add: Boolean) {
    val cloned = hologram.location.clone()
    for (line in hologram.lines) {
        when (line.type) {
            LineImpl.Type.TEXT_LINE, LineImpl.Type.CLICKABLE_TEXT_LINE -> {
                val tL = line as TextLine

                // add to lines
                tL.setCurrentLocation(cloned.clone())

                if (add) {
//                    hologram.lines.add(0, tL)
                } else {
                    hologram.pvt.seeingPlayers.forEach { tL.teleport(it) }
                }
                cloned.z += 0.175 * tL.value.length
            }

            else -> throw RuntimeException("This method load supports only TextLine & TextALine & ClickableTextLine.")
        }
    }
}